Singleton S3Client

jkinkead opened the pull request:
Fixes #16230 .

This drastically speeds up performance of interactions with S3, and eliminates a lot of spurious log warnings when interacting with S3 files.

The filesystem unit tests went from taking ~40 seconds to taking ~4 seconds with this change, a 10X performance improvement.

Some items of note:
- I updated the delete test to work on a bucket that had tests run previously. Without this change, a manual wipe of the file in quest was required after each run.
- I moved the request timeout to a central location, instead of being local to the `Sync` operation.
- I eliminated the increased connection timeout for `Sync`, which shouldn't be needed.
- Configuration is no longer a static variable protected by a mutex, but instead created as-needed. This should be non-functional, given that config is only created once during normal operation now.

jkinkead committed:
Ensure that the DeleteFile test works without special setup.

jkinkead committed:
Only create one S3Client for the lifetime of the filesystem.

drpngx commented on tensorflow/core/platform/s3/s3_file_system_test.cc:
```diff
@@ -130,6 +130,8 @@ TEST_F(S3FileSystemTest, NewReadOnlyMemoryRegionFromFile) {
 TEST_F(S3FileSystemTest, FileExists) {
   const string fname = TmpDir("FileExists");
+  // Ensure the file doesn't yet exist.
+  TF_ASSERT_OK(s3fs.DeleteFile(fname));
```
Isn't DeleteFile of a file that doesn't exist supposed to return an error?

jkinkead replied:
Good question! [The API docs](https://github.com/tensorflow/tensorflow/blob/04b5c75aae4bdbdac7c713714a369f9b360daf70/tensorflow/core/platform/file_system.h#L150-L151) don't specify.

The posix filesystem looks like it will return ENOENT if the file doesn't exist, but the S3 filesystem currently treats this as success. I'd rather not change the S3 behavior, but I could future-proof this by removing the `TF_ASSERT_OK` - what do you think?

drpngx replied:
Actually, I think we should return an error if the file doesn't exist. That's the expectation, and if we have different behavior for filesystems, then it will lead to workarounds at the call site.

jkinkead replied:
I'm worried about making a breaking change to the API here; it also worries me that it's out-of-scope for the main fix.

I can make it if you like, but it feels like it should be tracked separately.

jkinkead replied:
Also, from what I can tell, S3 doesn't communicate the delete status back from a delete. We would need to add a check for the object existing prior to deleting (fine, but not ideal).

drpngx commented on tensorflow/core/platform/s3/s3_file_system.h:
```diff
@@ -16,11 +16,14 @@ limitations under the License.
 namespace tensorflow {
 
 class S3FileSystem : public FileSystem {
+ private:
```
put after public section

drpngx commented on tensorflow/core/platform/s3/s3_file_system.h:
```diff
@@ -16,11 +16,14 @@ limitations under the License.
 
 class S3FileSystem : public FileSystem {
+ private:
+  std::shared_ptr<Aws::S3::S3Client> s3Client_;
```
I'd rather we use the google style guide: `s3_client_`.

Why is it a shared ptr? Can it be a `std::unique_ptr`?

jkinkead committed:
Rename variables to match Google C++ style.

Move private section to bottom of class definition.

jkinkead reviewed (commented):
Comments addressed!

rmlarsen commented:
@drpngx can you take another look, please?

drpngx commented:
Sounds reasonable.

jkinkead commented:
It looks like the failure is legit; there's a test that's asserting that stderr matches a specific string, and it's failing because of items printed out by the AWS client initialization.

This exposes a problem with greedy loading of the client: We'll see errors printed out in environments without valid credentials on startup. I'll make the client lazily-initialized, which should fix the test, and keep the errors from showing up unless you're trying to use S3.

drpngx commented:
I'm not sure we want these messages spilled out every time we start tensorflow.

drpngx commented:
Were these introduced in this PR?

jkinkead committed:
Initialize S3 client lazily.

Register a deleter to handle cleanup instead of always cleaning up in
the filesystem destructor.

jkinkead commented:
Formerly, the client was being initialized with each call to the S3 filesystem - so the messages were being printed iff someone accessed S3, but were also printed *each time* someone accessed S3.

I updated this to initialize lazily, once. The python tests should pass now, but I couldn't verify locally (I was getting an error about an undefined `enum` module trying to run them).

jkinkead committed:
Merge remote-tracking branch 'upstream/master' into s3_cleanup

rmlarsen commented:
@drpngx do you approve this?
